test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -10151,6 +10151,47 @@ BOOST_AUTO_TEST_CASE(constant_string)
 	ABI_CHECK(callContractFunction("h()"), encodeDyn(string("hello")));
 }
 
+BOOST_AUTO_TEST_CASE(abi_encode)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function f() returns (bytes) {
+				return abi.encode(1, 2);
+			}
+			function g() returns (bytes) {
+				return abi.encodePacked(uint256(1), uint256(2));
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(u256(32), u256(64), u256(1), u256(2)));
+	ABI_CHECK(callContractFunction("g()"), encodeArgs(u256(32), u256(64), u256(1), u256(2)));
+}
+
+BOOST_AUTO_TEST_CASE(abi_encode_call)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function f(uint8 a, string b) {
+				require(a == 42);
+				require(b.length == 2);
+				require(bytes(b)[0] == 72); // 'H'
+				require(bytes(b)[1] == 101); // 'e'
+			}
+			function g(uint8 a) {
+				bytes request = abi.encode(
+					bytes4(keccak256(abi.encodePacked("f(uint8,string)"))),
```

Comment:
This made me realise we could offer `abi.encodeSelector(string)` as a shorthand for this. Though having it a library seems better, too bad `abi` cannot be extended by libraries (yet).

Reply:
Yes, we do want that, but perhaps `encodeWithSelector`

Reply:
There are two options. I was going with:
```
abi.encode(
  abi.encodeSelector("f(uint8,string)"),
  a,
  "He"
);
```

Otherwise we can have three functions with varargs:
- `abi.encode`
- `abi.encodeWithSelector` (should enforce the first parameter being a string type ...)
- `abi.encodePacked`

Reply:
Unfortunately it does need the third version, otherwise it would be packed.

This means for `encodeWithSelector` we need to enforce the first parameter to be `string` (to be hashed) or `bytes4` (already hashed).